vispy/scene/widgets/grid.py

```diff
@@ -183,63 +258,242 @@ def __repr__(self):
+
+        Grid._add_gridding_width_constraints(self._solver, self._width_grid)
+        Grid._add_gridding_height_constraints(self._solver, self._height_grid)
+
+        Grid._add_stretch_constraints(self._solver,
+                                      self._width_grid,
+                                      self._height_grid,
+                                      self._grid_widgets,
+                                      self._widget_grid)
+
+        Grid._add_widget_dim_constraints(self._solver,
+                                         self._width_grid,
+                                         self._height_grid,
+                                         self._var_w,
+                                         self._var_h,
+                                         self._grid_widgets)
+
+    def _update_child_widget_dim(self):
+        # think in terms of (x, y). (row, col) makes code harder to read
+        ymax, xmax = self.grid_size
+        if ymax <= 0 or xmax <= 0:
+            return
+
+        rect = self.rect  # .padded(self.padding + self.margin)
+        if rect.width <= 0 or rect.height <= 0:
+            return
+        try:
+            if self._need_solver_recreate:
+                self._need_solver_recreate = False
+                self._recreate_solver()
+
+            # yes, this little dance is necessary for cassowary
+            # to not screw up :/
+            if self._height_stay:
+                self._solver.remove_constraint(self._height_stay)
+
+            self._var_h.value = rect.height
+            self._height_stay = self._solver.add_stay(self._var_h,
+                                                      strength=REQUIRED)
+
+            # self._var_w.value = rect.width
+            if self._width_stay:
+                self._solver.remove_constraint(self._width_stay)
+
+            self._var_w.value = rect.width
+            self._width_stay = self._solver.add_stay(self._var_w,
+                                                     strength=REQUIRED)
+
+            value_vectorized = np.vectorize(lambda x: x.value)
+
+            for (_, val) in self._grid_widgets.items():
+                (row, col, rspan, cspan, widget) = val
+
+                width = np.sum(value_vectorized(
+                               self._width_grid[row][col:col+cspan]))
+                height = np.sum(value_vectorized(
+                                self._height_grid[col][row:row+rspan]))
+                if col == 0:
+                    x = 0
+                else:
+                    x = np.sum(value_vectorized(self._width_grid[row][0:col]))
+
+                if row == 0:
+                    y = 0
+                else:
+                    y = np.sum(value_vectorized(self._height_grid[col][0:row]))
+
+                widget.size = (width, height)
+                widget.pos = (x, y)
+
+        except RequiredFailure as e:
+                print("cassowary failure\n%s" % e)
```

Comment:
We probably don't really want to print the errors, we should catch them and do something intelligent with them. We could probably just say in the docs that un-solvable systems will lead to no change in layout...?